test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -89,110 +106,162 @@ final case class Spec[+L, +T](caseValue: SpecCase[L, T, Spec[L, T]]) { self =>
-  final def foreachParN[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Parallel)(f)
+  final def foreachParN[R1 <: R, E1, A](
+    n: Int
+  )(failure: E => ZIO[R1, E1, A], success: T => ZIO[R1, E1, A]): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.ParallelN(n))(failure, success)
 
   /**
    * Returns a new spec with remapped labels.
    */
-  final def mapLabel[L1](f: L => L1): Spec[L1, T] =
-    transform[L1, T] {
+  final def mapLabel[L1](f: L => L1): Spec[R, E, L1, T] =
+    transform[R, E, L1, T] {
       case SuiteCase(label, specs, exec) => SuiteCase(f(label), specs, exec)
       case TestCase(label, test)         => TestCase(f(label), test)
     }
 
   /**
    * Returns a new spec with remapped tests.
    */
-  final def mapTest[T1](f: T => T1): Spec[L, T1] =
-    transform[L, T1] {
+  final def mapTest[T1](f: T => T1): Spec[R, E, L, T1] =
+    transform[R, E, L, T1] {
       case SuiteCase(label, specs, exec) => SuiteCase(label, specs, exec)
-      case TestCase(label, test)         => TestCase(label, f(test))
+      case TestCase(label, test)         => TestCase(label, test.map(f))
     }
 
   /**
-   * Computes the size of the spec, i.e. the number of tests in the spec.
+   * Uses the specified `Managed` to provide each test in this spec with its
+   * required environment.
    */
-  final def size: Int = fold[Int] {
-    case SuiteCase(_, counts, _) => counts.sum
-    case TestCase(_, _)          => 1
+  final def provideManaged[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] =
+    transform[Any, E1, L, T] {
+      case SuiteCase(label, specs, exec) => SuiteCase(label, specs.provideManaged(managed), exec)
+      case TestCase(label, test)         => TestCase(label, test.provideManaged(managed))
+    }
+
+  /**
+   * Uses the specified `Managed` once to provide all tests in this spec with
+   * a shared version of their required environment. This is useful when the
+   * act of creating the environment is expensive and should only be performed
+   * once.
+   */
+  final def provideManagedSuite[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] = {
+    def loop(r: R)(spec: Spec[R, E, L, T]): ZIO[Any, E, Spec[Any, E, L, T]] =
+      spec.caseValue match {
+        case SuiteCase(label, specs, exec) =>
+          specs.flatMap(ZIO.foreach(_)(loop(r))).map(z => Spec.suite(label, ZIO.succeed(z.toVector), exec)).provide(r)
+        case TestCase(label, test) =>
+          test.map(t => Spec.test(label, ZIO.succeed(t))).provide(r)
+      }
+    caseValue match {
+      case SuiteCase(label, specs, exec) =>
+        Spec.suite(label, managed.use(r => specs.flatMap(ZIO.foreach(_)(loop(r))).map(_.toVector).provide(r)), exec)
+      case TestCase(label, test) =>
+        Spec.test(label, test.provideManaged(managed))
+    }
   }
 
+  /**
+   * Computes the size of the spec, i.e. the number of tests in the spec.
+   */
+  final def size: ZIO[R, E, Int] =
```

Comment:
I wonder just how useful all these become in an effectful world. Basically computing the size here could, in theory, allocate and de-allocate resources required only to run the tests!

Reply:
I totally agree. I wonder if we move these to some type of syntax class for executed Specs or require implicit evidence that the spec has been executed. The only variants of these `fold` type methods we use are `exists` and `forall` in checking if all executed specs passed.